### PR TITLE
Optimize memory usage by adjusting the config size of exchange layer

### LINKED ISSUE #5924: Consider setting CHIP_CONFIG_RMP_RETRANS_TABLE_SIZE to the min of pool size and exchange context count
#### Problem
Right now if we have a limited number or possible exchange contexts but a large packetbuffer pool, we can end up with a large CRMP retransmit table.

 #### Proposed Solution
Take both values into account when defining the retransmit table size.

@yufengwangca Do you have time for this?

### PR DESCRIPTION
**Pull request by yufengwangca**
#### Problem
1. Right now if we have a limited number or possible exchange contexts but a large packetbuffer pool, we can end up with a large CRMP retransmit table.

2.  MCSP cache pool size is defined as CHIP_CONFIG_MAX_EXCHANGE_CONTEXTS, but we need to reserve two additional exchanges, one for MCSP protocol itself and another one for a temporary exchange to ack If we didn't find an existing exchange that matches the received message, and no unsolicited message handler registered to hand this message.

 #### Summary of Changes
1. Take both pool size and CHIP_CONFIG_MAX_EXCHANGE_CONTEXTS into account when defining the retransmit table size.
2. Reserve two extra exchanges, one for MCSP messages and another one for temporary exchange for ack.

 Fixes #5924

### CODE REVIEW

**Review by msandstedt — APPROVED** [review]

File: src/messaging/MessageCounterSync.h [hunk 1]
```diff
@@ -128,11 +128,15 @@ class MessageCounterSyncMgr : public Messaging::ExchangeDelegate
 
     Messaging::ExchangeManager * mExchangeMgr; // [READ ONLY] Associated Exchange Manager object.
 
-    // MessageCounterSyncProtocol cache table to queue the outging messages that trigger message counter synchronization protocol.
-    RetransTableEntry mRetransTable[CHIP_CONFIG_MAX_EXCHANGE_CONTEXTS];
-
-    // MessageCounterSyncProtocol cache table to queue the incoming messages that trigger message counter synchronization protocol.
-    ReceiveTableEntry mReceiveTable[CHIP_CONFIG_MAX_EXCHANGE_CONTEXTS];
+    // MessageCounterSyncProtocol cache table to queue the outging messages that trigger message counter
```
Comment (kpschoedel) [inline src/messaging/MessageCounterSync.h:131]: nit: ‘outgoing’

File: src/messaging/ReliableMessageProtocolConfig.h [hunk 2]
```diff
@@ -30,6 +30,10 @@
 namespace chip {
 namespace Messaging {
 
+#ifndef MIN
+#define MIN(a, b) (((a) < (b)) ? (a) : (b))
+#endif
```
Comment (kpschoedel) [inline src/messaging/ReliableMessageProtocolConfig.h:35]: nit: there's a generic chip::min() in CodeUtils.h (not sure why we don't use std::min(), though).

**Review by kpschoedel — APPROVED** [review]

**Review by bzbarsky-apple — APPROVED** [review]
Looks reasonable with @kpschoedel's issues addressed.

File: src/messaging/MessageCounterSync.h (hunk 1 above)
Comment (yufengwangca) [reply src/messaging/MessageCounterSync.h]: acked

File: src/messaging/ReliableMessageProtocolConfig.h (hunk 2 above)
Comment (yufengwangca) [reply src/messaging/ReliableMessageProtocolConfig.h]: acked

**Comment by yufengwangca** [comment]
@Damian-Nordic ?

**Review by Damian-Nordic — APPROVED** [review]

**Pull request merged**